### PR TITLE
Add option to throw memory-related exceptions from recompiler

### PR DESCRIPTION
**Pull request by Kingcom**
In some cases it can be useful to know when a game triggers invalid behavior, like unaligned memory access or TLB misses, that would crash a real system. This pull request adds a new option `Cpu.Recompiler.ThrowAddressExceptions` to achieve that. It can only be edited through the .ini file and is disabled by default.

TLB miss  exceptions were already thrown in developer builds, this is extended to when the new option is set. Loads and stores didn't check for unaligned access, so these checks are added. They are only written into the recompiled code if the option is enabled. Furthermore, as you explicitly opt-in to receive these exceptions, you definitely want to notice them - so any exceptions thrown due to the option are treated as immediately fatal.

### CODE REVIEW

File: pcsx2/Config.h [hunk 1]
```diff
@@ -259,6 +259,10 @@ struct Pcsx2Config
 				PreBlockCheckIOP:1;
 			bool
 				EnableEECache   :1;
+
+			bool
+				ThrowAddressExceptions :1;
+
```
Comment (RedDevilus) [inline pcsx2/Config.h:265]: Whitespaces

File: pcsx2/Pcsx2Config.cpp [hunk 2]
```diff
@@ -192,6 +194,8 @@ void Pcsx2Config::RecompilerOptions::LoadSave( IniInterface& ini )
 	IniBitBool( StackFrameChecks );
 	IniBitBool( PreBlockCheckEE );
 	IniBitBool( PreBlockCheckIOP );
+
```
Comment (RedDevilus) [inline pcsx2/Pcsx2Config.cpp:197]: Whitespace here as well

**Comment by tellowkrinkle** [comment]
Would you mind running `git clang-format HEAD~~` instead (which will only format your new changes)?  Clang format has issues with some of our macros (e.g. `jNO_DEFAULT` is a case and shouldn't be indented) so I'd rather not blindly run it on whole files right now

**Comment by Kingcom** [comment]
Sure. Formatting the files was suggested in Discord due to the inconsistencies already inside them, but I can definitely see how it would be a problem. I've rebased any clang-format changes on my changes into the commits. All the new lines should conform now, regardless of what's around them.

File: pcsx2/x86/ix86-32/iR5900LoadStore.cpp [hunk 3]
```diff
@@ -221,6 +292,21 @@ void recStore(u32 bits)
                 if (bits == 128)
                         xAND(arg1regd, ~0x0F);
 
+				if (EmuConfig.Cpu.Recompiler.ThrowAddressExceptions && bits != 128)
+				{
+					xTEST(arg1regd, alignment - 1);
+					xForwardJZ32 skipException;
```
Comment (tellowkrinkle) [inline pcsx2/x86/ix86-32/iR5900LoadStore.cpp:298]: This should be able to be a J8

File: pcsx2/x86/ix86-32/iR5900LoadStore.cpp [hunk 4]
```diff
@@ -221,6 +292,21 @@ void recStore(u32 bits)
...
+
+					// flush PC for the error message, preserve calculated address
+					xPUSH(arg1regd);
+					iFlushCall(FLUSH_PC);
+					xPOP(arg1regd);
```
Comment (tellowkrinkle) [inline pcsx2/x86/ix86-32/iR5900LoadStore.cpp:303]: What's the register save/restore here for?

**Review by tellowkrinkle — COMMENTED** [review]
I don't think you need to save registers around iFlushCall, which just writes things to memory (which on x86 doesn't need anything in registers)

**Comment by lightningterror** [comment]
Status on this?

**Comment by lightningterror** [comment]
Closing as it's not currently being worked on, can be reopened when it does.

**Pull request closed without merging**